transaction/CycleDetector.cpp

```diff
@@ -0,0 +1,122 @@
+          std::make_unique<StronglyConnectedComponents>(*wait_for_graph)) {
+}
+
+std::vector<DirectedGraph::node_id> CycleDetector::breakCycle() {
+  std::vector<DirectedGraph::node_id> nodes_to_kill;
+  const std::unordered_map<std::uint64_t, std::vector<DirectedGraph::node_id>>
+    component_mapping = strongly_connected_components_->getComponentMapping();
+  for (const std::pair<std::uint64_t, std::vector<DirectedGraph::node_id>>
+           &entry : component_mapping) {
+    // One node means no cycle.
+    if (entry.second.size() == 1) {
+      continue;
+    }
+    const std::vector<DirectedGraph::node_id> nodes
+        = breakComponent(entry.second);
+      for (const DirectedGraph::node_id &node : nodes) {
+        nodes_to_kill.push_back(node);
+      }
+  }
+  return nodes_to_kill;
+}
+
+std::vector<DirectedGraph::node_id>
+CycleDetector::breakComponent(const std::vector<DirectedGraph::node_id> &nodes) {
+  std::vector<DirectedGraph::node_id> targets;
+  // Convert it to set to ensure defensively that the elements are unique.
+  std::unordered_set<DirectedGraph::node_id> nodes_set(nodes.begin(), nodes.end());
+
+  while (true) {
+    const bool has_cycle = hasCycleWithin(nodes_set);
+    if (!has_cycle) {
+      break;
+    }
+    // If there is a cycle, start to pop a node from randomly.
+    // TODO(Hakan): This is very inefficient scheme, however in the
+    //              future, we can use the transaction's priority
+    //              as the victim selection parameter.
+    const DirectedGraph::node_id victim = *(nodes_set.begin());
+    nodes_set.erase(nodes_set.begin());
+    targets.push_back(victim);
+  }
+  return targets;
+}
+
+bool CycleDetector::hasCycleWithin(
+    const std::unordered_set<DirectedGraph::node_id> &within) {
+  // Keeps track of the nodes the algorithms visited.
+  std::unordered_set<DirectedGraph::node_id> visited;
+  for (const DirectedGraph::node_id node_id : within) {
+    // If it is visited, then pass to the next one.
+    if (visited.count(node_id) == 1) {
+      continue;
+    }
+    // Save the backtracking information.
+    std::stack<DirectedGraph::node_id> to_visit;
+    // Mark this id as "to be visited".
+    to_visit.push(node_id);
+    // Start to visit nodes until it is done.
+    while (!to_visit.empty()) {
+      const DirectedGraph::node_id current_node = to_visit.top();
+      to_visit.pop();
+      // Mark the node coming from stack as "visited".
+      visited.insert(current_node);
+      // For all adjacent nodes of this "visited" node,
+      const std::vector<DirectedGraph::node_id> adjacents
+          = wait_for_graph_->getAdjacentNodes(current_node);
+      for (DirectedGraph::node_id adj : adjacents) {
+        if (visited.count(adj) == 1) {
+          // If this adjacent node is a node we already visited, then
+          // there is a cycle.
+          return true;
+        } else if (within.count(adj) == 1 && visited.count(adj) == 0) {
```

Comment:
It won't be the case. Sometimes it is `within.count(adj) == 0` , and it means iterator is right now, does not point the node which is in first set (SCC). Then there is no point to continue with this path. That's why sometimes this for loop won't do anything.